src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1803,188 +1803,270 @@ public static bool Decompose(Matrix4x4 matrix, out Vector3 scale, out Quaternion
         {
             bool result = true;
 
-            unsafe
+            float det;
+
+            Matrix4x4 matTemp = Identity;
+            Vector3 canonicalBasis_Row0 = Unsafe.As<float, Vector3>(ref matTemp.M11);
+            Vector3 canonicalBasis_Row1 = Unsafe.As<float, Vector3>(ref matTemp.M21);
+            Vector3 canonicalBasis_Row2 = Unsafe.As<float, Vector3>(ref matTemp.M31);
+            Vector3 pVectorBasis_0 = Unsafe.As<float, Vector3>(ref matrix.M11);
+            Vector3 pVectorBasis_1 = Unsafe.As<float, Vector3>(ref matrix.M21);
+            Vector3 pVectorBasis_2 = Unsafe.As<float, Vector3>(ref matrix.M31);
```

Comment:
Reading this as a Vector4 and setting `W` to zero would probably be faster (single load, rather than 2, and can efficiently use `insertps` to zero).